### PR TITLE
NVMeFC Discover Login Functions

### PR DESCRIPTION
**Pull request by harshitap26**
# Description
A few sentences describing the overall goals of the pull request's commits.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
|https://github.com/dell/csm/issues/260|

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [ ] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [x] Tested on linux environment with nvme library installed
- [x] NVMe/FC Discovery and Login functions are successful

image

For the Ports that are linked (4 NVMe Ports), discovery is successful and hence connected with 2 fc_ports each.
image

### CODE REVIEW

**Review by harishp8889 — APPROVED** [review]

File: gonvme_fc.go [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+func (nvme *NVMeFC) nvmeDisconnect(target NVMeTarget) error {
+	// nvme disconnect is done via the nvme cli
+	// nvme disconnect -n <target NQN>
+	exe := nvme.buildNVMeCommand([]string{NVMeCommand, "disconnect", "-n", target.TargetNqn})
+	cmd := exec.Command(exe[0], exe[1:]...)
+
+	_, err := cmd.Output()
+
+	if err != nil {
+		fmt.Printf("\nError logging %s at %s: %v", target.TargetNqn, target.Portal, err)
```
Comment (spriya-m) [inline gonvme_fc.go:339]: error message can be specific to disconnect

File: gonvme_fc.go [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+
+	if len(nqns) == 0 {
+		return nqns, err
+	}
+
+	return nqns, nil
+}
+
+// DiscoverNVMeTargets - runs NVMe/FC discovery and returns a list of targets.
+func (nvme *NVMeFC) DiscoverNVMeTargets(targetAddress string, login bool) ([]NVMeTarget, error) {
```
Comment (spriya-m) [inline gonvme_fc.go:141]: Why do we need two functions each? DiscoverNVMeTargets and discoverNVMeTargets, GetInitiators() and getInitiators(), etc?

File: gonvme_fc.go [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+
+			tokens := strings.Fields(line)
+			if len(tokens) < 2 {
+				continue
+			}
+			key := tokens[0]
+			value := strings.Join(tokens[1:], "")
+			switch key {
+
+			case "=====Discovery":
```
Comment (spriya-m) [inline gonvme_fc.go:209]: please fix indentation.

File: gonvme_fc.go [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+			targets = append(targets, nvmeTarget)
+		}
+	}
+
+	if len(targets) == 0 {
+		fmt.Printf("\n Error discovering NVMe/FC targets: %s", err)
+		return []NVMeTarget{}, err
+	}
+
+	// TODO: Add optional login
```
Comment (spriya-m) [inline gonvme_fc.go:271]: remove commented code

File: gonvme_fc.go [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+	_, err := cmd.Output()
+
+	if err != nil {
+		if exiterr, ok := err.(*exec.ExitError); ok {
+			// nvme connect exited with an exit code != 0
+			nvmeConnectResult := -1
+			if status, ok := exiterr.Sys().(syscall.WaitStatus); ok {
+				nvmeConnectResult = status.ExitStatus()
+			}
+			if nvmeConnectResult == 114 || nvmeConnectResult == 70 {
```
Comment (spriya-m) [inline gonvme_fc.go:302]: are these error codes specific to any OS/platform? both codes denote session already exists?

**Review by spriya-m — COMMENTED** [review]
Also, go.sum is missing from the repo. can you please check that in?

File: gonvme_fc.go (hunk 5 above)
Comment (harshitap26) [reply gonvme_fc.go:302]: Yes, the error codes are different for nvme connect in different platforms

File: gonvme_fc.go (hunk 3 above)
Comment (harshitap26) [reply gonvme_fc.go:211]: The indentation changes back to this on running the gofmt command

Comment (spriya-m) [reply gonvme_fc.go:209]: ok

**Comment by harshitap26** [comment]
> Also, go.sum is missing from the repo. can you please check that in?

go.sum has been added in this PR

**Pull request closed without merging**